[SPARK-21146] [CORE] Master/Worker should handle and shutdown when any thread gets UncaughtException

devaraj-kavali opened the pull request:
## What changes were proposed in this pull request?

Adding the default UncaughtExceptionHandler to the Worker.

## How was this patch tested?

I verified it manually, when any of the worker thread gets uncaught exceptions then the default UncaughtExceptionHandler will handle those exceptions.

Commit:
[SPARK-21146] [CORE] Worker should handle and shutdown when any thread
gets UncaughtException

AmplabJenkins commented:
Can one of the admins verify this patch?

jiangxb1987 commented:
Could you elaborate on when we need this improvement? In fact I don't see much benefit of having this at first glance.

devaraj-kavali commented:
@jiangxb1987 Thanks for looking at this.

Any one of the threads in Worker gets an exception which is unhandled then the thread gets terminated and process(Worker) keeps running, and the Worker doesn't perform the terminated thread functionality. One instance I mentioned in the JIRA description, where dispatcher-event thread got the exception and terminated, worker keeps running without performing the terminated thread functionality. I think we should handle this at the process level for all the threads, if any thread gets unhandled exception then the process(Worker) should exit instead of performing partial functionality. 

 This PR code change would set the default UncaughtExceptionHandler as SparkUncaughtExceptionHandler and for any unhandled exception in any thread of the process, SparkUncaughtExceptionHandler handles the exception and exit the process with the appropriate exit code.

jiangxb1987 commented:
AFAIK SparkUncaughtExceptionHandler is designed to make Executors fail fast, because Executors are relatively lightweight, but I'm not sure we can afford to let Workers, even the Master fail fast too.
cc @cloud-fan for more comments.

cloud-fan commented:
according to the JIRA, if the worker meets some unrecoverable issues and the state becomes DEAD, it seems make sense to terminate this worker as it's not functional. But can you double check that this change won't terminate a still-functional worker?

devaraj-kavali commented:
> But can you double check that this change won't terminate a still-functional worker?

The change doesn't impact/harm the fully functional Worker, SparkUncaughtExceptionHandler handles only when any thread gets terminated due to the exception thrown from it without handling.

zsxwing commented:
ok to test

zsxwing commented:
LGTM. I thought we already did it :(

SparkQA commented:
**[Test build #78404 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78404/testReport)** for PR 18357 at commit [`c6e71f8`](https://github.com/apache/spark/commit/c6e71f87b0e8adbeadf35be76072c833bdcb3331).

SparkQA commented:
**[Test build #78404 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78404/testReport)** for PR 18357 at commit [`c6e71f8`](https://github.com/apache/spark/commit/c6e71f87b0e8adbeadf35be76072c833bdcb3331).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78404/
Test PASSed.

devaraj-kavali commented:
@zsxwing Thanks for looking into this, I see SparkUncaughtExceptionHandler exits the process for all the exceptions/errors.

Can we modify SparkUncaughtExceptionHandler in such a way that it kills the process for all Errors(always) and kills the process for Exceptions only when the flag is true, this flag can be enabled for Executors and disabled for other daemons(Master, Worker and HistoryServer)?

zsxwing commented:
@devaraj-kavali SparkUncaughtExceptionHandler is a singleton so you cannot get the configuration from SparkConf.

devaraj-kavali commented:
can we change SparkUncaughtExceptionHandler to class and create instance in each daemons main() with the constructor flag whether to kill the process for Exceptions or not?  Something like below,

For Executor : 
--------------
Thread.setDefaultUncaughtExceptionHandler(new SparkUncaughtExceptionHandler(true)) // Here true denotes to kill the process for Exceptions

For Other Daemons: 
--------------
Thread.setDefaultUncaughtExceptionHandler(new SparkUncaughtExceptionHandler(false)) // false denotes not to kill the process for Exceptions

Commit:
Updated SparkUncaughtExceptionHandler to take an argument whether to
kill/not-to-kill the process for the unhandled exceptions with the default
value as true, behaviour remains same for unhandled Error instances.

SparkQA commented:
**[Test build #78838 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78838/testReport)** for PR 18357 at commit [`a2f30f0`](https://github.com/apache/spark/commit/a2f30f07e7e72dfacb801032bfc71bb9fda1b650).

SparkQA commented:
**[Test build #78838 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78838/testReport)** for PR 18357 at commit [`a2f30f0`](https://github.com/apache/spark/commit/a2f30f07e7e72dfacb801032bfc71bb9fda1b650).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78838/
Test FAILed.

jiangxb1987 commented:
retest this please.

SparkQA commented:
**[Test build #78912 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78912/testReport)** for PR 18357 at commit [`a2f30f0`](https://github.com/apache/spark/commit/a2f30f07e7e72dfacb801032bfc71bb9fda1b650).

SparkQA commented:
**[Test build #78912 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78912/testReport)** for PR 18357 at commit [`a2f30f0`](https://github.com/apache/spark/commit/a2f30f07e7e72dfacb801032bfc71bb9fda1b650).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78912/
Test PASSed.

zsxwing commented on core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:
```diff
@@ -26,27 +26,34 @@ import org.apache.spark.internal.Logging
  */
 private[spark] object SparkUncaughtExceptionHandler
   extends Thread.UncaughtExceptionHandler with Logging {
+  private[this] var exitOnException = true
```
Instead of using a static variable, I prefer to change `SparkUncaughtExceptionHandler` to a class.

zsxwing commented on core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:
```diff
@@ -26,27 +26,34 @@ import org.apache.spark.internal.Logging
+        case t: Throwable => Runtime.getRuntime.halt(SparkExitCode.UNCAUGHT_EXCEPTION_TWICE)
       }
-    } catch {
-      case oom: OutOfMemoryError => Runtime.getRuntime.halt(SparkExitCode.OOM)
```
Could you keep these codes? It unlikely happens but since the codes are there, it's better to not change it.

devaraj-kavali replied:
@zsxwing, this code is still there but I moved the try&catch to the block where we invoke System.exit. Do you mean moving the whole code in uncaughtException() to try block and having the catch block?


 ```
+      } catch {
 +        case oom: OutOfMemoryError => Runtime.getRuntime.halt(SparkExitCode.OOM)
 +        case t: Throwable => Runtime.getRuntime.halt(SparkExitCode.UNCAUGHT_EXCEPTION_TWICE)
```

zsxwing replied:
> Do you mean moving the whole code in uncaughtException() to try block and having the catch block?

Yes.

devaraj-kavali replied:
@zsxwing Thanks for the clarification.

Commit:
Fixed the review comments

SparkQA commented:
**[Test build #78997 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78997/testReport)** for PR 18357 at commit [`56c2e4d`](https://github.com/apache/spark/commit/56c2e4d1ff167db531bec4a3e5ff15fb412fdfbc).

SparkQA commented:
**[Test build #78997 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78997/testReport)** for PR 18357 at commit [`56c2e4d`](https://github.com/apache/spark/commit/56c2e4d1ff167db531bec4a3e5ff15fb412fdfbc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78997/
Test PASSed.

zsxwing commented on core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:
```diff
@@ -20,29 +20,29 @@ package org.apache.spark.util
         if (exception.isInstanceOf[OutOfMemoryError]) {
           System.exit(SparkExitCode.OOM)
         } else {
           System.exit(SparkExitCode.UNCAUGHT_EXCEPTION)
```
The current changes are too much. Could you rename `exitOnException` to `exitOnUncaughtException` and just change this line to
```
if (exitOnUncaughtException) {
  System.exit(SparkExitCode.UNCAUGHT_EXCEPTION)
}
```

zsxwing commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -737,6 +737,7 @@ private[deploy] object Worker extends Logging {
   val ENDPOINT_NAME = "Worker"
 
   def main(argStrings: Array[String]) {
+    Thread.setDefaultUncaughtExceptionHandler(new SparkUncaughtExceptionHandler(false))
```
nit: `false` => `exitOnException = false`

zsxwing commented on core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:
```diff
@@ -20,29 +20,29 @@ package org.apache.spark.util
+ * any Errors, and also terminates the process for Exceptions when the exitOnException flag is true.
  */
-private[spark] object SparkUncaughtExceptionHandler
+private[spark] class SparkUncaughtExceptionHandler(val exitOnException: Boolean = true)
```
nit: please also document `exitOnException` using

```
@param exitOnException ...
```

zsxwing commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -1037,6 +1037,7 @@ private[deploy] object Master extends Logging {
   val ENDPOINT_NAME = "Master"
 
   def main(argStrings: Array[String]) {
+    Thread.setDefaultUncaughtExceptionHandler(new SparkUncaughtExceptionHandler(false))
```
nit: `false` => `exitOnException = false`

Commit:
Fixed the review comments

SparkQA commented:
**[Test build #79298 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79298/testReport)** for PR 18357 at commit [`3d5106b`](https://github.com/apache/spark/commit/3d5106b4b0ec67350e89b1b7579cd2c164bc1b4b).

zsxwing commented:
LGTM. Pending tests. Thanks!

SparkQA commented:
**[Test build #79298 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79298/testReport)** for PR 18357 at commit [`3d5106b`](https://github.com/apache/spark/commit/3d5106b4b0ec67350e89b1b7579cd2c164bc1b4b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79298/
Test PASSed.

devaraj-kavali commented:
@zsxwing Tests have passed, can you check this? Thanks

zsxwing commented:
@devaraj-kavali Sorray. I forgot this PR. I will trigger a new run as master has been updated a lot. Will set a reminder for me to merge this PR :)

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #79537 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79537/testReport)** for PR 18357 at commit [`3d5106b`](https://github.com/apache/spark/commit/3d5106b4b0ec67350e89b1b7579cd2c164bc1b4b).

SparkQA commented:
**[Test build #79537 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79537/testReport)** for PR 18357 at commit [`3d5106b`](https://github.com/apache/spark/commit/3d5106b4b0ec67350e89b1b7579cd2c164bc1b4b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79537/
Test FAILed.

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #79542 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79542/testReport)** for PR 18357 at commit [`3d5106b`](https://github.com/apache/spark/commit/3d5106b4b0ec67350e89b1b7579cd2c164bc1b4b).

SparkQA commented:
**[Test build #79542 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79542/testReport)** for PR 18357 at commit [`3d5106b`](https://github.com/apache/spark/commit/3d5106b4b0ec67350e89b1b7579cd2c164bc1b4b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79542/
Test PASSed.

zsxwing commented:
Thanks! Merging to master.
